### ISSUE TITLE
[ENH] add `.fillna` method to `ListMethods`

### ISSUE THREAD

**Issue by thomcom** [rapidsai/cudf#13274]
**Describe the bug**
.fillna does not work with `ListColumns`.

**Steps/Code to reproduce bug**
```
x = cudf.Series([0, 1, None])
```
```
x.fillna(5)
0    0
1    1
2    5
dtype: int64
```
```
y = cudf.Series([[0, 1], [2, 3], [None, None]])
y.fillna(5) 
```
```
0    [0.0, 1.0]
1    [2.0, 3.0]
2    [nan, nan]
dtype: list
```

**Expected behavior**
`2    [nan, nan]` should be `2    [5, 5]`

**Environment overview (please complete the following information)**
cudf 23.06 installed via pip

**Environment details**
Not available.

**Additional context**

**Comment by thomcom** [comment]
This is needed because `GeoSeries` in cuSpatial need to be able to replace `NA` with `np.nan` in order to easily print them. Ref https://github.com/rapidsai/cuspatial/issues/1102

**Comment by thomcom** [comment]
The fix appears to be
```
new_children = [child for child in y._column.points._column._base_children[:-1]]
new_children.append(y._column.points._column._base_children[-1].fillna(np.nan))
y._column.points._column.set_base_children(tuple(new_children))
```

I don't have a working cudf build at the moment but I'll look into a PR.

**Comment by wence-** [comment]
I think this is working as designed. `NA` or `NaN` for empty is a top-level flag indicating that the relevant entry in the series is missing. On the other hand, in your case you want to dive into a list and replace a non-top-level flag. But I don't think this is right: `None` is not the same as `[None]`.

Certainly making this change would be an API break from pandas:
```python
import pandas as pd

x = pd.Series([[1], [2], [pd.NA]])
y = x.fillna(5)
x == y # => True

x = pd.Series([1, 2, pd.NA])
y = x.fillna(5)
x == y # => False
```

**Comment by shwina** [comment]
I think the API we want to expose for this is `s.list.fillna()`.

**Comment by wence-** [comment]
> I think the API we want to expose for this is `s.list.fillna()`.

I guess that traverses arbitrarily deep (since if you have nested lists the `ListMethods` object doesn't expose the intermediate levels, I think)

**Comment by thomcom** [comment]
Thanks for the follow up on this @wence- I certainly don't want to break any pandas compatibility. I'm writing `s.list.fillna()` now.

**Comment by thomcom** [comment]
Yeah I'm hesitant, do you really want a `.list.fillna` method that traverses the arbitrary depth of the `ListColumn` and replaces all of the null values? That's where I am in development now, do you want me to finish it and send a PR? @shwina @wence-

**Comment by thomcom** [comment]
I could just write this for cuspatial and leave cudf alone. :)

**Comment by thomcom** [comment]
Additionally, `cudf` seems very sensitive about what kinds of `ListColumn` can be created from the `Series` constructor. Arbitrary nesting seems allowed if the only value is `None`, if I mix `None` with floating points I get funky errors.

```
s = cudf.Series([
    [[[[None]]]],
    [[[None]]],
    [[None]],
    [None],
])
s
```
```
0    [[[[None]]]]
1      [[[None]]]
2        [[None]]
3          [None]
dtype: list
```

but
```
s = cudf.Series([
    [[[[None]]]],
    [[[None]]],
    [[None]],
    [None, 0.0],
])
s
```

gets
```
ArrowInvalid: cannot mix list and non-list, non-null values
```

so I'm less certain that handling complex nesting cases is worth pursuing. Opinions?

**Comment by wence-** [comment]
Oh wow, that's an odd one, that seems like an arrow thing? The way I _thought_ lists in cudf work is that every rows must have the same type, so your first entry has type `List (List (List (List (int | None))))`. Now, _I_ think that `List None` does not have the same type as `None`, and so I would have thought that writing `[[None], None]` would not be allowed because the first entry has type `List None` and the second has type `None`. However, it seems like arrow treats `None` as an arbitrary placeholder at any depth which blows a hole through the type system in my book.

I suppose the type is something like:

```
data List a = None | [a | None]
```

**Comment by shwina** [comment]
That's right, at any level of nesting, you can have a "null" list. So `[[1, 2, 3], None, [1, 2, None]]` is a valid list of lists in cuDF/Arrow.

**Comment by isVoid** [comment]
```
s = cudf.Series([
    [[[[None]]]],
    [[[None]]],
    [[None]],
    [None, 0.0],
])
s
```

This fails because [[[[None]]]] and [None, 0.0] isn't the same type. As @wence- points out, the first row is `List (List (List (List (int | None))))`, the second is `List( float | None )`.

**Comment by vyasr** [comment]
I'm going to close this since it seems like `Series.fillna` is already behaving as expected and in #13317 we could not come to a decision on suitable semantics for `Series.list.fillna` and there was also no pressing need outside of cuspatial (which is now discontinued). We can reopen if a future need arises.
